src/main/java/com/iota/iri/service/validation/impl/TransactionSolidifierImpl.java

```diff
@@ -0,0 +1,383 @@
+        if(transactionViewModel.getApprovers(tangle).size() == 0) {
+            tipsViewModel.addTipHash(transactionViewModel.getHash());
+        }
+        tipsViewModel.removeTipHash(transactionViewModel.getTrunkTransactionHash());
+        tipsViewModel.removeTipHash(transactionViewModel.getBranchTransactionHash());
+
+        if(quickSetSolid(transactionViewModel)) {
+            transactionViewModel.update(tangle, snapshotProvider.getInitialSnapshot(), "solid|height");
+            tipsViewModel.setSolid(transactionViewModel.getHash());
+            addToPropagationQueue(transactionViewModel.getHash());
+        }
+    }
+
+    @Override
+    public void addToPropagationQueue(Hash hash) throws Exception{
+        if(!solidTransactions.contains(hash)) {
+            if (solidTransactions.size() >= MAX_SIZE) {
+                solidTransactions.poll();
+            }
+            solidTransactions.put(hash);
+        }
+    }
+
+    @Override
+    public boolean quickSetSolid(final TransactionViewModel transactionViewModel) throws Exception {
+        if(!transactionViewModel.isSolid()) {
+            boolean solid = true;
+            if (!checkApproovee(transactionViewModel.getTrunkTransaction(tangle))) {
+                solid = false;
+            }
+            if (!checkApproovee(transactionViewModel.getBranchTransaction(tangle))) {
+                solid = false;
+            }
+            if(solid) {
+                transactionViewModel.updateSolid(true);
+                transactionViewModel.updateHeights(tangle, snapshotProvider.getInitialSnapshot());
+                addToPropagationQueue(transactionViewModel.getHash());
+                addToBroadcastQueue(transactionViewModel);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * If the the {@code approvee} is missing, request it from a neighbor.
+     * @param approovee transaction we check.
+     * @return true if {@code approvee} is solid.
+     * @throws Exception if we encounter an error while requesting a transaction
+     */
+    private boolean checkApproovee(TransactionViewModel approovee) throws Exception {
+        if(snapshotProvider.getInitialSnapshot().hasSolidEntryPoint(approovee.getHash())) {
+            return true;
+        }
+        if(approovee.getType() == PREFILLED_SLOT) {
+            // don't solidify from the bottom until cuckoo filters can identify where we deleted -> otherwise we will
+            // continue requesting old transactions forever
+            //transactionRequester.requestTransaction(approovee.getHash(), false);
+            return false;
+        }
+        return approovee.isSolid();
+    }
+
+    @VisibleForTesting
+    void propagateSolidTransactions() {
+        while(!Thread.currentThread().isInterrupted() && solidTransactions.peek() != null) {
+            try {
+                Hash hash = solidTransactions.poll();
```

Comment:
I have a theory on the solidification problem...
I think we are trying to solidify before we go through the receive stage

So we fail to `getApprovers` because they were not received yet stored
And when they come in, the non received transaction is polled out already...

Makes sense?